File: server/app/schemas/command.py
import uuid

from pydantic import BaseModel, Field


class SlashCommand(BaseModel):
    """One command as advertised by the bots service, for the composer's `/` popup."""

    name: str
    description: str
    # Argument hint shown next to the name, e.g. "<url ou busca>". Empty when
    # the command takes none.
    usage: str = ""
    # Whether invoking it requires the caller to be in a voice channel. The
    # bots service enforces it; this is only so the client can say why upfront.
    requires_voice: bool = False


class RunCommandRequest(BaseModel):
    """What the composer sends when a message turns out to be a slash command."""

    channel_id: uuid.UUID
    name: str = Field(min_length=1, max_length=32, pattern=r"^[a-z0-9_-]+$")
    args: str = Field(default="", max_length=2000)

File: app/lib/models/command.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'command.freezed.dart';
part 'command.g.dart';

/// One slash command the server's bots answer to, as offered in the composer's
/// `/` menu. Port of the `SlashCommand` in `lib/types.ts`.
///
/// The list comes from `GET /api/commands` and is empty on a deployment with no
/// bots service — then the menu simply never opens.
@freezed
abstract class SlashCommand with _$SlashCommand {
  const factory SlashCommand({
    required String name,
    required String description,

    /// Argument hint shown next to the name, e.g. `<url ou busca>`.
    @Default('') String usage,

    /// The bot enforces this; the client only uses it to explain the
    /// requirement before spending a round trip.
    @Default(false) bool requiresVoice,
  }) = _SlashCommand;

  factory SlashCommand.fromJson(Map<String, dynamic> json) => _$SlashCommandFromJson(json);
}
